From c904b4e905ac8a0a0c885e931c34ae26928e4ab5 Mon Sep 17 00:00:00 2001
From: Marek Marczykowski <marmarek@invisiblethingslab.com>
Date: Sun, 28 Apr 2013 01:12:49 +0200
Subject: [PATCH] libxl: do not assume Dom0 backend while listing disks and
 nics
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Organization: Invisible Things Lab
Cc: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

One more place where code assumed that all backends are in dom0. List
devices in domain device/ tree, instead of backend/ of dom0.
Additionally fix libxl_devid_to_device_{nic,disk} to fill backend_domid
properly.

Signed-off-by: Marek Marczykowski <marmarek@invisiblethingslab.com>
Signed-off-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

Ported to Xen 4.7.0 (marcus@wetwa.re)
--- xen-4.7.0/tools/libxl/libxl.c.orig	2016-07-15 20:07:49.913000000 +0300
+++ xen-4.7.0/tools/libxl/libxl.c	2016-07-15 22:02:42.927000000 +0300
@@ -3593,6 +3593,13 @@
 
     libxl_device_nic_init(nic);
 
+    rc = sscanf(libxl_path, "/local/domain/%d/", &nic->backend_domid);
+    if (rc != 1) {
+        rc = ERROR_FAIL;
+        LOG(ERROR, "Unable to fetch device backend domid from %s", libxl_path);
+        goto out;
+    }
+
     tmp = READ_LIBXLDEV(gc, "handle");
     if (tmp)
         nic->devid = atoi(tmp);
